### PR TITLE
Add VarZeroVecFormat support to VarTuple and make_var

### PR DESCRIPTION
**Pull request by Manishearth**
Part of holistic work in https://github.com/unicode-org/icu4x/issues/5523

Part of https://github.com/unicode-org/icu4x/issues/2312 (Somehow never filed a separate issue for this, but I mentioned it in a commetn)

### CODE REVIEW

File: utils/zerovec/derive/src/make_varule.rs [hunk 1]
```diff
@@ -488,21 +488,29 @@ struct UnsizedField<'a> {
...
 }
 
 impl<'a> UnsizedFields<'a> {
-    fn new(fields: Vec<UnsizedField<'a>>) -> Self {
+    /// The format_param is an optional tokenstream describing a VZVFormat argument needed by MultiFieldsULE
+    fn new(fields: Vec<UnsizedField<'a>>, format_param: Option<TokenStream2>) -> Self {
         assert!(!fields.is_empty(), "Must have at least one unsized field");
-        Self { fields }
+
+        let format_param = format_param.unwrap_or_else(|| quote!(zerovec::vecs::Index16));
```
Comment (sffc) [inline utils/zerovec/derive/src/make_varule.rs:499]: Thought: make it a required parameter?

File: utils/zerovec/src/ule/multi.rs [hunk 2]
```diff
@@ -4,21 +4,20 @@
...
 ///
 /// Logically, consider it to be `(V1, V2, V3, ..)`
 /// where `V1` etc are potentially different [`VarULE`] types.
 ///
-/// Internally, it is represented by a VarZeroSlice.
+/// Internally, it is represented by a VarZeroSlice without the length part.
 #[derive(PartialEq, Eq)]
 #[repr(transparent)]
-pub struct MultiFieldsULE<const LEN: usize, Format: VarZeroVecFormat = Index32>(
+pub struct MultiFieldsULE<const LEN: usize, Format: VarZeroVecFormat>(
```
Comment (sffc) [inline utils/zerovec/src/ule/multi.rs:20]: Suggestion: Make sure all the safety comments say that the `Format` needs to be the same between validation and reading fields. It seems like a footgun that the MultiFieldsULE could be potentially deserialized with one format and read with another format. Making the format be explicit is probably a positive change.

File: utils/zerovec/src/ule/tuplevar.rs [hunk 3]
```diff
@@ -30,17 +31,17 @@ macro_rules! tuple_varule {
         #[doc = concat!("VarULE type for tuples with ", $len, " elements. See module docs for more information")]
         #[repr(transparent)]
         #[allow(clippy::exhaustive_structs)] // stable
-        pub struct $name<$($T: ?Sized),+> {
+        pub struct $name<$($T: ?Sized,)+ Format: VarZeroVecFormat = Index16> {
```
Comment (sffc) [inline utils/zerovec/src/ule/tuplevar.rs:34]: Suggestion: make the format explicit?

I'm OK with the default for higher-level types, but for a lower-level type, let's just make everyone pick the right value? And this is a highly unsafe type, so being explicit seems safer.

File: utils/zerovec/src/ule/tuplevar.rs [hunk 4]
```diff
@@ -267,22 +271,29 @@ mod tests {
...
+                Tuple3VarULE<str, [u8], VarZeroSlice<str>, Format>,
             >(val);
         }
     }
+
+    #[test]
+    fn test_tripleule_validate() {
+        test_tripleule_validate_inner::<Index8>();
+        test_tripleule_validate_inner::<Index16>();
+        test_tripleule_validate_inner::<Index32>();
```
Comment (sffc) [inline utils/zerovec/src/ule/tuplevar.rs:297]: Suggestion: add a test that serializing and deserializing between different formats is usually an error (unless we're unlucky and the bits are valid in both formats)

File: utils/zerovec/src/ule/tuplevar.rs (hunk 3 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/tuplevar.rs:34]: Why is this a highly unsafe type?

Comment (Manishearth) [reply utils/zerovec/src/ule/tuplevar.rs:34]: You mean impl-wise? I don't think the format shows up too often in the unsafe code here.

I'd rather not have most users care about the format. I don't think this is too low-level, it's a tuple.

Comment (sffc) [reply utils/zerovec/src/ule/tuplevar.rs:34]: ok

File: utils/zerovec/src/ule/multi.rs (hunk 2 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/multi.rs:20]: I updated the comments to explicitly mention the format where it talks about VZLS. I also made VZLS require the format.

File: utils/zerovec/src/ule/tuplevar.rs (hunk 4 above)
Comment (Manishearth) [reply utils/zerovec/src/ule/tuplevar.rs:305]: I don't see the point of a test like that. If it succeeds, it's a valid representation. We already have scattered tests that ensure that invalid types don't parse with carefully chosen non-parsing data, I don't see any benefit to writing one around swapping formats specifically.

**Review by sffc — APPROVED** [review]

**Pull request merged**